packages/client/lib/rpc/modules/eth.ts

```diff
@@ -39,24 +77,174 @@ export class Eth {
+      [validators.blockOption],
+    ])
+
+    this.getTransactionCount = middleware(this.getTransactionCount.bind(this), 2, [
+      [validators.address],
+      [validators.blockOption],
+    ])
+
     this.protocolVersion = middleware(this.protocolVersion.bind(this), 0, [])
   }
 
   /**
    * Returns number of the most recent block.
-   * @param  {Array<*>} [params] An empty array
-   * @return {Promise}
+   * @param params An empty array
    */
   async blockNumber(_params = []) {
     const latestHeader = await this._chain.getLatestHeader()
     return `0x${latestHeader.number.toString(16)}`
   }
 
   /**
-   * Returns information about a block by block number
-   * @param  {Array<BN|bool>} [params] An array of two parameters: An big integer of a block number and a
-   * boolean determining whether it returns full transaction objects or just the transaction hashes
-   * @return {Promise}
+   * Executes a new message call immediately without creating a transaction on the block chain.
+   * Currently only "latest" block number is supported.
+   * @param params An array of two parameters:
+   *   1. The transaction object
+   *       * from (optional) - The address the transaction is sent from
+   *       * to - The address the transaction is directed to
+   *       * gas (optional) - Integer of the gas provided for the transaction execution
+   *       * gasPrice (optional) - Integer of the gasPrice used for each paid gas
+   *       * value (optional) - Integer of the value sent with this transaction
+   *       * data (optional) - Hash of the method signature and encoded parameters.
+   *   2. integer block number, or the string "latest", "earliest" or "pending"
+   * @returns The return value of the executed contract.
+   */
+  async call(params: [RpcCallTx, string]) {
+    const [transaction, blockOpt] = params
+
+    const latestBlockNumber = await this.blockNumber()
+    if (blockOpt !== 'latest' && blockOpt !== latestBlockNumber) {
+      // todo: this can be resolved with some kind of functionality of stateAt(blockNumber)
+      return {
+        code: INVALID_PARAMS,
+        message: `Currently only block option "latest" supported`,
+      }
+    }
+
+    if (!this._vm) {
+      throw new Error('missing vm')
+    }
+
+    if (!transaction.gas) {
+      // If no gas limit is specified use the last block gas limit as an upper bound.
+      const latestHeader = await this._chain.getLatestHeader()
+      transaction.gas = latestHeader.gasLimit as any
+    }
+
+    const vm = this._vm.copy()
+    const txData = { ...transaction, gasLimit: transaction.gas }
+    const tx = Transaction.fromTxData(txData, { common: vm._common, freeze: false })
+
+    // set from address
+    const from = transaction.from ? Address.fromString(transaction.from) : Address.zero()
+    tx.getSenderAddress = () => {
+      return from
+    }
+
+    const { execResult } = await vm.runTx({ tx })
+    return bufferToHex(unpadBuffer(execResult.returnValue))
```

Comment:
is `unpadBuffer` here correct? otherwise the test contract [`returns (address addr)`](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1130/files#diff-6f5214b59c0bdda424d9c3634ff8af4d09e8bfdb28aa0b67af962e625c18434fR33) return value comes back as 32 bytes instead of 20.

without `unpadBuffer` it would return here: `0x000000000000000000000000ccfd725760a68823ff1e062f4cc97e1360e8d997`

Reply:
I am 99% sure that we don't want to unpad it. We want to return the actual return value. The ABI handlers will then decode the raw buffer. For instance, if ABI knows it is an address which is returned, then it will expect 32 bytes as output and will only return the last 20 bytes.

Reply:
ah that makes sense, ok i will update, thanks!

Reply:
updated in https://github.com/ethereumjs/ethereumjs-monorepo/pull/1130/commits/6a48e434664ddc8a6c9597cd4fb8697864f1c921